### PR TITLE
prov/gni: fixes for use of fi_inject

### LINKED ISSUE #559: problem with fi_inject and friends for certain communication patterns
It turns out that `fi_inject` and friends can cause problems for the GNI provider for certain communication patterns.  The **osu_barrier** test run with 4 endpoints (MPI processes) illustrates this well.  The barrier timing part works fine, the problem is with the `MPI_Reduce` to get the timing statistics.  With an unmodified GNI provider which has an inject size of 64 bytes, the test hangs.  With an inject size of 0, the test passes.  The problem has to do with the fact that the communication pattern at this job size differs between the barrier and reduce patterns.  Note the reduce is a one-way operation for some ranks, they just send to a parent.  In the case of the Open MPI OFI MTL, `fi_tinject` is used.  If the VC connection from the sender to the receiver endpoint is not yet set up, the sender ranks end up in `MPI_Finalize` in an out-of-band barrier.  The VC never got moved to completed state. 

If the gni provider is modified to advertise an inject size of 0, the test passes.

There may be other osu collectives that have a similar behavior pattern.

### PR DESCRIPTION
**Pull request by hppritcha**
fi_inject could fail for certain edge cases
in Open MPI since at finalize, the MPI processes
fall in to an out-of-band barrier that prevents
completion of inject requests.  This commit fixes
the issue.

@ztiffany 
@sungeunchoi 

It might be best for cases where control progress is set to `FI_CONTROL_MANUAL` by the application that we set the max inject size for the GNI provider to 0.

Fixes ofi-cray/libfabric-cray#559
Signed-off-by: Howard Pritchard howardp@lanl.gov

### CODE REVIEW

**Comment by sungeunchoi** [comment]
Can we get a gnitest to test this?

**Comment by sungeunchoi** [comment]
Whether we set the inject size to zero or leave it, I think we should document the behavior for `FI_CONTROL_MANUAL`.

**Comment by hppritcha** [comment]
I'll see about a gni test.  I'll do a second PR after this one to set inject size to zero if control is manual.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
@sungeunchoi  added a criterion check.

File: prov/gni/test/rdm_sr.c [hunk 1]
```diff
@@ -736,6 +737,54 @@ Test(rdm_sr, inject)
...
+	sz = fi_recv(ep[1], target, len, rem_mr[1], gni_addr[0], source);
+	cr_assert_eq(sz, 0);
+
+	/*
+	 * do progress until send counter is updated.
+	 * This works because we have FI_PROGRESS_AUTO for control progress
+	 */
+	while (fi_cntr_read(send_cntr[0]) < 1) {
+		pthread_yield();
+	}
```
Comment (sungeunchoi) [inline prov/gni/test/rdm_sr.c:767]: I thought the original bug had NO calls to libfabric functions after the inject, hence no opportunity to make the connection.  This seems like it would work without your fix, or am I misunderstanding something?

Comment (hppritcha) [inline prov/gni/test/rdm_sr.c:767]: It should fail since progessing the send_cntr[0] doesn't progress the domain[1]/ep[1] connection setup if there is no control progress thread for domain[1].    I'll check that.

**Comment by hppritcha** [comment]
I verified that with FI_CONTROL_MANUAL the test case hangs.
